full-node/sov-stf-runner/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+pub struct StateTransitionRunner<ST, DA, Vm>
+where
+    DA: DaService,
+    Vm: Zkvm,
+    ST: StateTransitionFunction<Vm, <<DA as DaService>::Spec as DaSpec>::BlobTransaction>,
+{
+    start_height: u64,
+    da_service: DA,
+    app: ST,
+    ledger_db: LedgerDB,
+    state_root: StateRoot<ST, Vm, DA>,
+    listen_address: SocketAddr,
+}
+
+impl<ST, DA, Vm> StateTransitionRunner<ST, DA, Vm>
+where
+    DA: DaService<Error = anyhow::Error> + Clone + Send + Sync + 'static,
+    Vm: Zkvm,
+    ST: StateTransitionFunction<Vm, <<DA as DaService>::Spec as DaSpec>::BlobTransaction>,
+{
+    /// Creates a new `StateTransitionRunner` runner.
+    pub fn new(
+        rollup_config: RollupConfig,
+        da_service: DA,
+        ledger_db: LedgerDB,
+        mut app: ST,
+        is_storage_empty: bool,
+        genesis_config: InitialState<ST, Vm, DA>,
+    ) -> Result<Self, anyhow::Error> {
+        let rpc_config = rollup_config.rpc_config;
+
+        let prev_state_root = {
+            // Check if the rollup has previously been initialized
+            if is_storage_empty {
+                info!("No history detected. Initializing chain...");
+                app.init_chain(genesis_config);
+                info!("Chain initialization is done.");
+            } else {
+                debug!("Chain is already initialized. Skipping initialization.");
+            }
+
+            let res = app.apply_slot(Default::default(), []);
+            // HACK: Tell the rollup that you're running an empty DA layer block so that it will return the latest state root.
+            // This will be removed shortly.
+            res.state_root
+        };
+
+        let listen_address = SocketAddr::new(rpc_config.bind_host.parse()?, rpc_config.bind_port);
+
+        // Start the main rollup loop
+        let item_numbers = ledger_db.get_next_items_numbers();
+        let last_slot_processed_before_shutdown = item_numbers.slot_number - 1;
+        let start_height = rollup_config.start_height + last_slot_processed_before_shutdown;
+
+        Ok(Self {
+            start_height,
+            da_service,
+            app,
+            ledger_db,
+            state_root: prev_state_root,
+            listen_address,
+        })
+    }
+
+    /// Starts an rpc server with provided rpc methods.
+    pub async fn start_rpc_server(&self, methods: RpcModule<()>) {
+        let listen_address = self.listen_address;
+        let _handle = tokio::spawn(async move {
+            let server = jsonrpsee::server::ServerBuilder::default()
+                .build([listen_address].as_ref())
+                .await
+                .unwrap();
+
+            info!("Starting RPC server at {} ", server.local_addr().unwrap());
+            let _server_handle = server.start(methods).unwrap();
+            futures::future::pending::<()>().await;
+        });
+    }
+
+    /// Runs the rollup.
+    pub async fn run(&mut self) -> Result<(), anyhow::Error> {
```

Comment:
It is out of scope of this PR, but how stf runner shuts down? 

I guess it can do graceful shutdown after processing slot.

Reply:
Now it just runs, let's implement shutdown in another PR.